pkg/identityserver/email.go

```diff
@@ -0,0 +1,129 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package identityserver
+
+import (
+	"context"
+	"net/url"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/jinzhu/gorm"
+	"go.thethings.network/lorawan-stack/pkg/email"
+	"go.thethings.network/lorawan-stack/pkg/email/sendgrid"
+	"go.thethings.network/lorawan-stack/pkg/email/smtp"
+	"go.thethings.network/lorawan-stack/pkg/identityserver/emails"
+	"go.thethings.network/lorawan-stack/pkg/identityserver/store"
+	"go.thethings.network/lorawan-stack/pkg/log"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+)
+
+var globalEmailTemplateRegistry = email.NewTemplateRegistry(nil)
+
+func (is *IdentityServer) getEmailTemplates(_ context.Context) *email.TemplateRegistry {
+	// TODO: Get email template fetcher from config.
+	// Re-use existing registry for same fetcher (to avoid re-compiling templates).
+	return globalEmailTemplateRegistry
+}
+
+// SendEmail sends an email.
+func (is *IdentityServer) SendEmail(ctx context.Context, f func(emails.Data) email.MessageData) (err error) {
+	isConfig := is.configFromContext(ctx)
+	var sender email.Sender
+	switch isConfig.Email.Provider {
+	case "sendgrid":
+		sender, err = sendgrid.New(ctx, isConfig.Email.Config, isConfig.Email.SendGrid)
+	case "smtp":
+		sender, err = smtp.New(ctx, isConfig.Email.Config, isConfig.Email.SMTP)
+	}
+	if err != nil {
+		return err
+	}
+	var data emails.Data
+	data.Network.Name = isConfig.Email.Network.Name
+	identityServerURL, _ := url.Parse(isConfig.Email.Network.IdentityServerURL)
+	identityServerURL.Path = "/"
```

Comment:
This is incorrect. The URL in the config can be assumed to be the final URL to use.